File: game_logic.rb
require_relative "goban"
require_relative "sgf_reader"
require_relative "handicap_setter"

# GameLogic enforces the game logic.
class GameLogic
  attr_reader :goban, :komi, :cur_color, :game_ended, :game_ending, :who_resigned
  
  def initialize
    @console = false
    @history = []
    @errors = []
    @handicap = 0
    @who_resigned = nil
    @goban = nil
  end

  def new_game(size=nil, handicap=@handicap, komi=nil)
    @history.clear
    @errors.clear
    @num_pass = 0
    @cur_color = BLACK
    @game_ended = @game_ending = false
    @who_resigned = nil
    if ! @goban or ( size and size != @goban.size )
      @goban = Goban.new(size)
    else
      @goban.clear
    end
    @komi = (komi ? komi : (handicap == 0 ? 6.5 : 0.5))
    set_handicap(handicap)
  end
  
  # Initializes the handicap points
  # h can be a number or a string
  # string examples: "3" or "3=d4-p16-p4" or "d4-p16-p4"
  def set_handicap(h)
    raise "Handicap cannot be changed during a game" if @history.size > 0
    @handicap = HandicapSetter.set_handicap(@goban, h)
    # White first when handicap
    @cur_color = WHITE if @handicap != 0
    return true
  end

  # game is a series of moves, e.g. "c2,b2,pass,b4,b3,undo,b4,pass,b3"
  def load_moves(game)
    begin
      game = sgf_to_game(game)
      game.split(",").each do |move|
        if !play_one_move(move) then raise "Failed playing the loaded move: #{move}" end
      end
      return true
    rescue => err
      error_msg "Failed loading moves. Please double check the format of your input."
      error_msg "Error: #{err.message} (#{err.class.name})"
      $log.error("Error while loading moves:\n#{err}\n#{err.backtrace}")
      return false
    end
  end

  # Handles a regular move + the special commands (pass, resign, undo, load, hand, log)
  # Returns false if a problem occured. In this case the error message is available.
  def play_one_move(move)
    if @game_ended then return error_msg("Game already ended") end
    # $log.debug("GameLogic playing #{Grid.color_name(@cur_color)}: #{move}") if $debug
    if /^[a-z][1-2]?[0-9]$/ === move
      return play_a_stone(move)
    elsif move == "undo"
      return request_undo
    elsif move.start_with?("resi")
      return resign
    elsif move == "pass"
      return pass_one_move
    elsif move.start_with?("hand")
      return set_handicap(move.split(":")[1])
    elsif move.start_with?("load:")
      return load_moves(move[5..-1])
    elsif move.start_with?("log")
      return set_log_level(move.split(":")[1])
    else
      return error_msg "Invalid command: #{move}"
    end
  end

  # Handles a new stone move (not special commands like "pass")
  def play_a_stone(move)
    i, j = Grid.parse_move(move)
    if !Stone.valid_move?(@goban, i, j, @cur_color) then return error_msg("Invalid move: #{move}") end
    Stone.play_at(@goban, i, j, @cur_color)
    store_move_in_history(move)
    next_player!
    @num_pass = 0
    return true
  end
  
  # One player resigns.
  def resign
    store_move_in_history("resign")
    @who_resigned = @cur_color
    @game_ended = true
    return true
  end

  # Call this when the current player wants to pass.
  # If all (remaining) players pass, we go into "ending mode".
  # Caller is responsible of checking the GameLogic#game_ending flag:
  # If the flag goes to true, the method accept_ending (below) should be called next.
  def pass_one_move
    store_move_in_history("pass")
    @num_pass += 1
    @game_ending = true if @num_pass >= 2
    next_player!
    return true
  end
  
  # Call this each time GameLogic#game_ending goes to true (ending mode).
  # The score should be counted and proposed to players.
  # "accept" parameter should be true if all players accept the proposed ending (score count).
  # Only after this call will the game be really finished.
  # If accept=false, this means a player refuses to end here
  # => the game should continue until the next time all players pass.
  def accept_ending(accept)
    return error_msg "The game is not ending yet" if !@game_ending
    if !accept
      @game_ending = false # exit ending mode; we will play some more...
    else
      @game_ended = true # ending accepted. Game is finished.
    end
    return true
  end

  # Returns how many moves have been played so far
  # (can be bigger than the stone count since "pass" or "resign" are also moves)
  def move_number?
    return @history.size
  end

  # Returns a text representation of the list of moves played so far
  def history_string
    return (@handicap>0 ? "handicap:#{@handicap}," : "") +
      @history.join(",") +
      " (#{@history.size} moves)"
  end

  # Returns an array with the prisoner count per color
  # e.g. [3,5] means 3 black stones are prisoners, 5 white stones
  def prisoners?
    return Group.prisoners?(@goban)
  end
  
  # If called with on=true, error messages will be directly displayed on the console.
  # If not called, the default behavior needs the caller to use get_errors method.
  def messages_to_console(on=true)
    @console = on
  end

  # Returns the error messages noticed until now and clears the list.
  def get_errors
    errors = @errors.clone
    @errors.clear
    return errors
  end
  
  def set_log_level(cmd)
    begin
      a = cmd.split("=")
      flag = a[1].to_i != 0
      raise 0 if ! flag and a[1]!="0"
      case a[0]
      when "group" then $debug_group = flag
      when "ai" then $debug_ai = flag
      when "all" then $debug = $debug_group = $debug_ai = flag
      else raise 1
      end
      return true
    rescue
      return error_msg "Invalid log command: #{cmd}"
    end
  end

#===============================================================================
private
#===============================================================================

  def next_player!
    @cur_color = (@cur_color+1) % 2
  end

  # Always returns false
  def error_msg(msg)
    if ! @console then @errors.push(msg) else puts msg end
    return false
  end

  def store_move_in_history(move)
    @history.push(move)
  end
  
  # undo one full game turn (e.g. one black move and one white)
  def request_undo
    if @history.size < 2
      return error_msg "Nothing to undo"
    end
    2.times do
      Stone.undo(@goban) if ! @history.last.end_with?("pass") # no stone to remove for a pass
      @history.pop
    end
    @num_pass = 0
    return true
  end

  # Converts a game (list of moves) from SGF format to our internal format.
  # Returns the game unchanged if it is not an SGF one.
  # Returns an empty move list if nothing should be played (a game is pending).
  def sgf_to_game(game)
    return game if ! game.start_with?("(;FF") # are they are always the 1st characters?
    reader = SgfReader.new(game)
    new_game(reader.board_size, reader.handicap)
    @komi = reader.komi
    return reader.to_move_list
  end
  
end

File: js/GameLogic.js
'use strict';

var CONST = require('./constants');
var Goban = require('./Goban');
var Grid = require('./Grid');
var HandicapSetter = require('./HandicapSetter');
var log = require('./log');
var main = require('./main');
var SgfReader = require('./SgfReader');
var ruleConfig = require('../config/rules.json');

var BLACK = CONST.BLACK, WHITE = CONST.WHITE;

var DEFAULT_RULES = CONST.JP_RULES;


/** @class GameLogic enforces the game logic.
 * NB: it makes sense for GameLogic to keep the komi, e.g. in case AI want to evaluate 
 * the score in mid-game.
 * public read-only attribute: goban, komi, curColor, gameEnded, gameEnding, whoResigned
 */
function GameLogic(src) {
    this.history = [];
    this.errors = [];
    this.infos = {};
    this.playerNames = [];
    this.setRules(DEFAULT_RULES);

    this.goban = null;
    this.handicap = this.komi = this.numPass = 0;
    this.curColor = this.whoStarts = BLACK;
    this.gameEnding = this.gameEnded = null;
    this.whoResigned = this.resignReason = null;

    if (src) this.copy(src);
}
module.exports = GameLogic;


// Note that player info remains until this method is called again (i.e. for several games)
GameLogic.prototype.setPlayer = function (color, name) {
    this.playerNames[color] = name;
};

GameLogic.prototype.setRules = function (rulesName, okIfUnknown) {
    rulesName = rulesName || 'unspecified';
    this.rulesName = rulesName;
    var rules = ruleConfig['.' + rulesName.toLowerCase()];
    if (!rules) {
        if (!okIfUnknown) throw new Error('Invalid rules: ' + rulesName +
            '\nValid values: ' + Object.keys(ruleConfig).map(function (s) { return s.substr(1); }).join(', '));
        rules = ruleConfig['.' + DEFAULT_RULES.toLowerCase()];
    }
    this.usePositionalSuperko = rules.usePositionalSuperko || false;
    this.useAreaScoring = rules.useAreaScoring || false;
};

GameLogic.prototype.copy = function (src) {
    this.setWhoStarts(src.whoStarts);
    this.setPlayer(BLACK, src.playerNames[BLACK]);
    this.setPlayer(WHITE, src.playerNames[WHITE]);
    this.setRules(src.rulesName, /*okIfUnknown=*/true);

    this.newGame(src.goban.gsize, src.handicap, src.komi);

    this.loadMoves(src.history.join(','));
};

// Handicap and komi are optional.
// Returns true if size and handicap could be set to given values.
// NB: setRules must be called before this, unless rules did not change.
GameLogic.prototype.newGame = function (gsize, handicap, komi) {
    this.history.clear();
    this.errors.clear();
    this.numPass = 0;
    this.gameEnded = this.gameEnding = false;
    this.whoResigned = this.resignReason = null;

    if (!this.goban || gsize !== this.goban.gsize) {
        this.goban = new Goban(gsize);
    } else {
        this.goban.clear();
    }
    this.goban.setPositionalSuperko(this.usePositionalSuperko);

    handicap = handicap !== undefined ? handicap : 0;
    this.setHandicapAndWhoStarts(handicap);

    this.komi = komi !== undefined ? komi : (handicap ? 0.5 : 6.5);

    if (this.goban.gsize !== gsize) return this._errorMsg('Size could not be set: ' + gsize);
    if (this.handicap !== handicap) return this._errorMsg('Handicap could not be set: ' + handicap);
    return true;
};

GameLogic.prototype.setWhoStarts = function (color) {
    this.curColor = this.whoStarts = color;
};

// Initializes the handicap points
// h can be a number or a string
// string examples: "3" or "3=d4-p16-p4" or "d4-p16-p4"
GameLogic.prototype.setHandicapAndWhoStarts = function (h) {
    if (this.history.length > 0) {
        throw new Error('Handicap cannot be changed during a game');
    }
    this.handicap = HandicapSetter.setHandicap(this.goban, h);

    // White first when handicap > 0
    this.setWhoStarts(this.handicap > 0 ? WHITE : BLACK);
    return true;
};

GameLogic.prototype._failLoad = function (msg, errors) {
    if (log.error) log.error(msg);
    if (!errors) throw new Error(msg);
    errors.push(msg);
    return false;
};

// @param {string} game - moves, e.g. "c2,b2,pass,b4,b3,undo,b4,pass,b3"
// @param {string[]} [errors] - errors will be added to this or thrown
GameLogic.prototype.loadMoves = function (game, errors) {
    if (!game) return true;
    var moves = game.split(',');
    for (var i = 0; i < moves.length; i++) {
        if (!this.playOneMove(moves[i])) {
            return this._failLoad('Failed playing loaded move #' + (i + 1) + ':\n' +
                this.getErrors().join(', '), errors);
        }
    }
    return true;
};

/**
 + Loads an SGF format game
 * @param {string} game - SGF game text
 * @param {string[]} [errors] - errors will be added to this or thrown
 * @param {number} [upToMoveNumber] - loads moves up to the position before this - SGF only
 * @return {boolean} - true if succeeded
 */
GameLogic.prototype.loadSgf = function (game, errors, upToMoveNumber) {
    var reader = new SgfReader(), infos;
    try {
        this.infos = infos = reader.readGame(game, upToMoveNumber);
    } catch (err) {
        return this._failLoad('Failed loading SGF moves:\n' + err, errors);
    }

    this.newGame(infos.boardSize, 0, infos.komi); // handicap, if any, will be in move list
    this.setPlayer(BLACK, infos.playerBlack);
    this.setPlayer(WHITE, infos.playerWhite);
    this.setRules(infos.rules, /*okIfUnknown=*/true);

    return this.loadMoves(reader.toMoveList(), errors);
};

// Call this when the format of game to load is unknown
GameLogic.prototype.loadAnyGame = function (game, errors) {
    if (SgfReader.isSgf(game)) {
        return this.loadSgf(game, errors);
    } else {
        return this.loadMoves(game, errors);
    }
}; 

function getMoveColor(move) {
    switch (move[0]) {
    case 'B': return BLACK;
    case 'W': return WHITE;
    default: return undefined;
    }
}

GameLogic.prototype.stripMoveColor = function (move) {
    if (move[0] !== 'B' && move[0] !== 'W') return move;
    if (move[1] !== '-') return move.substr(1);
    return move.substr(2);
};

function addMoveColor(move, color) {
    return (color === BLACK ? 'B-' : 'W-') + move;
}

// Handles a regular move + the special commands (pass, resign, undo, load, hand, log)
// Returns false if a problem occured. In this case the error message is available.
GameLogic.prototype.playOneMove = function (move) {
    if (this.gameEnded) return this._errorMsg('Game already ended');

    if (/^[B|W]?-?[a-z][1-2]?[0-9]$/.test(move)) {
        return this.playOneStone(move);
    }
    var cmd = this.stripMoveColor(move);
    if (cmd.startsWith('pass')) {
        return this.passOneMove();
    } else if (cmd.startsWith('resi')) {
        return this.resign(getMoveColor(move));
    } else if (cmd === 'undo') {
        return this._requestUndo();
    } else if (cmd === 'half_undo') {
        return this._requestUndo(true);
    } else if (cmd.startsWith('hand')) {
        return this.setHandicapAndWhoStarts(cmd.split(':')[1]);
    } else if (cmd.startsWith('load:')) {
        return this.loadMoves(cmd.slice(5));
    } else if (cmd.startsWith('log')) {
        return this.setLogLevel(cmd.split(':')[1]);
    } else {
        return this._errorMsg('Invalid command: ' + cmd);
    }
};

// Handles a new stone move (not special commands like "pass")
// e.g. "c3" or "Bc3" or "Wc3"
GameLogic.prototype.playOneStone = function (move) {
    var coords = this.oneMove2xy(move);
    if (coords.length === 3) this.curColor = coords[2];
    var i = coords[0], j = coords[1];
    if (!this.goban.isValidMove(i, j, this.curColor)) {
        return this._errorMsg('Invalid move: ' + move);
    }
    this.goban.playAt(i, j, this.curColor);
    this._storeMoveInHistory(move);
    this._nextPlayer();
    this.numPass = 0;
    return true;
};

// Parses a move like "c12" into [3,12]; OR "Bc1" into [3,1,0]
// Returns null if move is "pass" or "resi(gn)"; throws if move is illegal
GameLogic.prototype.oneMove2xy = function (move) {
    var color = getMoveColor(move);
    if (color !== undefined) move = this.stripMoveColor(move);

    var coords = Grid.move2xy(move, /*dontThrow=*/true);
    if (!coords) {
        if (move === 'pass' || move.startsWith('resi')) return null;
        throw new Error('Illegal move parsed: ' + move);
    }
    if (color !== undefined) return coords.concat(color);
    return coords;
};

// One player resigns.
GameLogic.prototype.resign = function (color, reason) {
    color = color !== undefined ? color : this.curColor;
    this.whoResigned = color;
    this.resignReason = reason;
    var move = addMoveColor('resign', color) + (reason ? '-' + reason : '');
    this._storeMoveInHistory(move);
    this.gameEnded = true;
    this.gameEnding = false;
    return true;
};

// Call this when the current player wants to pass.
// If all (remaining) players pass, we go into "ending mode".
// Caller is responsible of checking the GameLogic#game_ending flag:
// If the flag goes to true, the method accept_ending (below) should be called next.
GameLogic.prototype.passOneMove = function () {
    this._storeMoveInHistory(addMoveColor('pass', this.curColor));
    this.numPass++;
    if (this.numPass >= 2) {
        this.gameEnding = true;
    }
    this._nextPlayer();
    return true;
};

// Call this each time GameLogic#game_ending goes to true (ending mode).
// The score should be counted and proposed to players.
// "accept" parameter should be true if all players accept the proposed ending (score count).
// Only after this call will the game be really finished.
// If accept=false, this means a player refuses to end here
// => the game should continue until the next time all players pass.
GameLogic.prototype.acceptEnding = function (accept, whoRefused) {
    if (!this.gameEnding) return this._errorMsg('The game is not ending yet');
    this.gameEnding = false;
    if (accept) {
        this.gameEnded = true; // ending accepted. Game is finished.
        return true;
    }
    // Score refused (in dispute)
    // if the player who refused just played, we give the turn back to him
    if (whoRefused !== this.curColor) {
        this.history.pop(); // remove last "pass"
        this._nextPlayer();
    }
    return true;
};

// Returns how many moves have been played so far
// (can be bigger than the stone count since "pass" or "resign" are also moves)
GameLogic.prototype.moveNumber = function () {
    return this.history.length;
};

// Returns a text representation of the list of moves played so far
GameLogic.prototype.historyString = function () {
    var hand = this.handicap > 0 ? 'handicap:' + this.handicap + ',' : '';
    var h = this.history;
    var color1 = h.length && h[0].length === 2 ? addMoveColor('', this.whoStarts) : '';
    return hand + color1 + h.join(',');
};

// Stores a new error message
// Always returns false.
GameLogic.prototype._errorMsg = function (msg) {
    this.errors.push(msg);
    return false;
};

// Returns the error messages noticed until now and clears the list.
GameLogic.prototype.getErrors = function () {
    var errors = this.errors;
    this.errors = [];
    return errors;
};

GameLogic.prototype.setLogLevel = function (cmd) {
    var args = cmd.split('=');
    var flag = parseInt(args[1]) !== 0;
    switch (args[0]) {
    case 'group':
        main.debugGroup = flag;
        break;
    case 'ai':
        main.debugAi = flag;
        break;
    case 'all':
        main.debug = main.debugGroup = main.debugAi = flag;
        break;
    default: 
        return this._errorMsg('Invalid log command: ' + cmd);
    }
    return true;
};

GameLogic.prototype._nextPlayer = function () {
    this.curColor = 1 - this.curColor;
};

GameLogic.prototype._storeMoveInHistory = function (move) {
    return this.history.push(move);
};

// undo one full game turn (e.g. one black move and one white)
GameLogic.prototype._requestUndo = function (halfMove) {
    var count = halfMove ? 1 : 2;
    if (this.history.length < count) {
        return this._errorMsg('Nothing to undo');
    }
    for (var i = count; i >= 1; i--) {
        var move = this.history.pop();
        if (!move.endsWith('pass')) this.goban.undo();
    }
    if (halfMove) this._nextPlayer();
    this.numPass = 0;
    return true;
};
